### ISSUE TITLE
Bullets are displaced by 1 tick

### ISSUE THREAD

**Issue by RicochetYT** [mastercomfig/team-comtress-2#638]
### Description
Title says it all. Bullets are displaced by 1 tick upon firing.
I'm *pretty sure* this also occurs even with building from this repo, if it doesn't, I didn't notice it get fixed.
https://github.com/ValveSoftware/source-sdk-2013/pull/442
https://github.com/ValveSoftware/Source-1-Games/issues/4829
[Sourcemod plugin that fixes the bug on the serverside (you need to update offsets for it to work)](https://forums.alliedmods.net/showthread.php?t=315405)

I thought it would be worth putting this here, though it is an issue affecting all Source 1 games, to my knowledge. I already opened an issue in the [ValveSoftware/Source-1-Games](https://github.com/ValveSoftware/Source-1-Games) repo, but I thought I would put it here as well, hoping it has some chance of being included in the next TF2 update!

### Checklist

- [X] None of the open *or* closed issues document this problem.
- [X] This is a bug, not a new idea.
- [X] This is objectively a bug, and not something that people commonly disagree on.
- [X] This issue occurs on the latest Team Comtress release, or the `master` branch.
- [X] This issue occurs on the latest vanilla TF2 update.

- [X] This bug is in the code of the game, not in the textures, sounds, or other resources.
- [X] This issue is listed on the [Source 1 official bugtracker](https://github.com/ValveSoftware/Source-1-Games/issues).
- [X] This issue occurs with DirectX level 100.
- [X] This issue occurs when using no custom content.
- [X] This issue occurs when using `start_tf2.bat` (or `start_server.bat`) with no custom launch options.
- [X] This issue occurs when using a listen server (using the `map` command ingame).
- [X] This issue occurs when using a dedicated server.
- [ ] This bug is in the internal procedures or tooling (e.g. the build process).

### Reproduction

1. Load into a game (LAN or dedicated, doesn't matter.)
2. Set host_timescale to a low value
3. Move sideways at a fast speed, use noclip to exaggerate the displacement of the bullet
4. Fire a shot and watch as it doesn't hit the enemy despite hovering over said enemies head (you can use external software to see tracers or whatever else can be used to show shots...)

### Screenshots/Videos
https://www.youtube.com/watch?v=VPT0-CKODNc

https://www.youtube.com/watch?v=pr4EZ06mrpQ

#### Client Settings

- *Server or client:* Both?
- *Operating system:* Any
- *Team Comtress version:* Any
- *Custom content:* N/A
- *Launch options:* N/A

**Comment by mastercoms** [comment]
Would you be able to confirm this bug is fixed if a possible patch was made for it?

**Comment by RicochetYT** [comment]
This bug was not fixed in live TF2, however, there is a Sourcemod plugin that is a bandaid fix for this issue.

If you're referring to TC2/tf2-patches, I have not tested whether or not this bug exists but I can almost guarantee it does. I will try testing it soon.

**Comment by mastercoms** [comment]
I mean that if a patch were made for this bug, would you be able to reliably test?

**Comment by RicochetYT** [comment]
Oh, yeah, it's pretty easy to test.

**Comment by RicochetYT** [comment]
To elaborate on this, this bug is very easily replicable and is a consistent, every time issue when moving at any speed, the degree to which it affects shooting is just exaggerated with higher velocities on the localplayer.

**Comment by mastercoms** [comment]
Can you please retest with 2.0.0-b1?

**Comment by RicochetYT** [comment]
Sure

**Comment by RicochetYT** [comment]
> Can you please retest with 2.0.0-b1?

No, this issue is not fixed on 2.0.0-b1, I just tested on a dedicated and LAN server.

**Comment by mastercoms** [comment]
Could you post a video of the bug then?

**Comment by mastercoms** [comment]
Could you record a video of it in TC2?

**Comment by mastercoms** [comment]
Please retest with 2.0.0-b2

**Comment by RicochetYT** [comment]
> Could you record a video of it in TC2?

Build 2.0.0-b1:

https://user-images.githubusercontent.com/36606403/232143908-de3c6946-c764-48c4-9e41-d41b50593c69.mp4

> Please retest with 2.0.0-b2

Build 2.0.0-b2:

https://user-images.githubusercontent.com/36606403/232145597-ebe8fe36-4da6-4344-b2b0-79832ff2bdd6.mp4



Sorry for the late reply, I've been busy today.

**Comment by RicochetYT** [comment]
If you could add a debug command (if it doesn't already exist) to show a client eye-angle trace on shot compared to server eye-angles on shot [(like in this video)](https://youtu.be/VYs0-lUwWas), it would probably make this easier to see.

**Comment by RicochetYT** [comment]
Oh and I just found a new bug introduced most likely in 2.0.0-b2, seems like some odd rubber banding when firing, but it's inconsistently reproducible.

https://user-images.githubusercontent.com/36606403/232147383-ffc22065-63b1-4fdf-a6e3-14e88e2bbf44.mp4

**Comment by mastercoms** [comment]
Thanks for testing, I will see what I can do

**Comment by RicochetYT** [comment]
How it should look for reference (live TF2, scope removed for visibility, fixed with the plugin linked in the issue)

https://user-images.githubusercontent.com/36606403/232262941-86992fdc-0ae3-40e8-a904-870fabc7804b.mp4

**Comment by mastercoms** [comment]
I think now that it's a much better fix to make sure movement is not 1 tick behind, so I will try to fix that instead

**Comment by RicochetYT** [comment]
If anyone was curious, this issue was (seemingly) fixed in Counter-Strike 2 (or more likely, Source 2 as a whole).
This is almost certainly not intentional or something that's impossible to fix.

**Comment by RicochetYT** [comment]
> If anyone was curious, this issue was (seemingly) fixed in Counter-Strike 2 (or more likely, Source 2 as a whole).
> This is almost certainly not intentional or something that's impossible to fix.

I say "seemingly" because the bullet tracer + impact particles and hole are still skewed by your velocity in the direction you're moving on the client, but the server handles it perfectly and doesn't skew where you shot.

**Comment by XutaxKamay** [comment]
> > If anyone was curious, this issue was (seemingly) fixed in Counter-Strike 2 (or more likely, Source 2 as a whole).
> > This is almost certainly not intentional or something that's impossible to fix.
> 
> I say "seemingly" because the bullet tracer + impact particles and hole are still skewed by your velocity in the direction you're moving on the client, but the server handles it perfectly and doesn't skew where you shot.

Yep exactly, if you're ready to read wall of text; please checkout the two comments here, this is my view on this https://github.com/ValveSoftware/source-sdk-2013/pull/442#issuecomment-1770868507
